merge.go

```diff
@@ -0,0 +1,599 @@
+				typ:   d.addLabel,
+				props: d.setProp,
+			}
+			res = append(res, op.writeQuery(cfg))
+		case createEdgeOp:
+			op := createEdge{
+				e:     d.edge,
+				id:    d.id,
+				typ:   d.addLabel,
+				props: d.setProp,
+			}
+			res = append(res, op.writeQuery(cfg))
+		}
+	}
+	return res, vars
+}
+
+type OperationQueue struct {
+	Ops  []string
+	vars []map[string]interface{}
+}
+
+func RunOperations(ctx *ls.Context, session *Session, tx neo4j.Transaction, ops OperationQueue) error {
+	for ix, op := range ops.Ops {
+		_, err := tx.Run(op, ops.vars[ix])
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func Merge(memGraph, dbGraph *lpg.Graph, dbGraphIds map[*lpg.Node]int64, dbEdges map[*lpg.Edge]int64, config Config) (*lpg.Graph, OperationQueue, error) {
+	memEntitiesMap := ls.GetEntityInfo(memGraph)
+	dbEntitiesMap := ls.GetEntityInfo(dbGraph)
+	findDBCounterpart := func(n *lpg.Node, schemaPV, idPV *ls.PropertyValue) *lpg.Node {
+		if ls.AsPropertyValue(n.GetProperty(ls.EntitySchemaTerm)).IsEqual(schemaPV) {
+			if ls.AsPropertyValue(n.GetProperty(ls.EntityIDTerm)).IsEqual(idPV) {
+				return n
+			}
+		}
+		return nil
+	}
+	memEntities := make([]*lpg.Node, len(memEntitiesMap))
+	dbEntities := make([]*lpg.Node, len(memEntitiesMap))
+	// align by index matching entity roots i.e, first entity in mem slice should match first entity in db slice
+	ix := 0
+	for n := range memEntitiesMap {
+		schemaPV := ls.AsPropertyValue(n.GetProperty(ls.EntitySchemaTerm))
+		eidPV := ls.AsPropertyValue(n.GetProperty(ls.EntityIDTerm))
+		memEntities[ix] = n
+		for db := range dbEntitiesMap {
+			if findDBCounterpart(db, schemaPV, eidPV) != nil {
+				dbEntities[ix] = db
+				delete(dbEntitiesMap, db)
+				continue
+			}
+		}
+		ix++
+	}
+	deltas := make([]delta, 0)
+	for ix := 0; ix < len(memEntities); ix++ {
+		d, _ := mergeEntity(memEntities[ix], dbEntities[ix], dbGraph, dbGraphIds, dbEdges)
+		deltas = append(deltas, d...)
+	}
+	queries, vars := buildQueriesFromDeltas(deltas, config)
+	opsQueue := OperationQueue{Ops: queries, vars: vars}
+	return dbGraph, opsQueue, nil
+}
+
+func (dx delta) setDeltaToNode(dbNode *lpg.Node) {
+	for _, v := range dx.removeProp {
+		dbNode.RemoveProperty(v)
+	}
+	for k, v := range dx.setProp {
+		dbNode.SetProperty(k, v)
+	}
+	if len(dx.addLabel) > 0 {
+		dbNode.SetLabels(lpg.NewStringSet(dx.addLabel...))
+	}
+}
+
+func (dx delta) setDeltaToEdge(dbEdge *lpg.Edge) {
+	for _, v := range dx.removeProp {
+		dbEdge.RemoveProperty(v)
+	}
+	for k, v := range dx.setProp {
+		dbEdge.SetProperty(k, v)
+	}
+	if len(dx.addLabel) > 0 {
+		dbEdge.SetLabel(strings.Join(dx.addLabel, ","))
+	}
+}
+
+type step struct {
+	edge *lpg.Edge
+	node *lpg.Node
+}
+
+type edgeStep struct {
+	to   *lpg.Node
+	from *lpg.Node
+	edge *lpg.Edge
+}
+
+func mergeSubtree(memNode, dbNode *lpg.Node, dbGraph *lpg.Graph, dbGraphIds map[*lpg.Node]int64, dbEdges map[*lpg.Edge]int64, isChild bool) []delta {
+	nodeAssociations := make(map[*lpg.Node][]step)
```

Comment:
Pass nodeAssociations and edgeAssociations as arg to mergeSubtree. These should be initialized in main Merge func()